### PR TITLE
[inductor] add CppBuilder.

### PR DESCRIPTION
**Pull request by xuhancn**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116178

POC link: https://github.com/xuhancn/x86_isa_help 
Change logs：
1. Use new CppBuilder which also support Windows MSVC.
2. Add cpuid based x86 isa detector, it is support all x86 machine and all OS. (replace "/proc/cpuinfo")
3. Switch CppCodeCache to new CppBuilder.
4. Switch AotCodeCache to new CppBuilder.
5. Use new CppBuilder manage CppWrapperCodeCache args.

TODO:
1. Make is_fbcode works.
2. cleanup code.

Potential work:
1. Use CppBuilder to instead of cpp_extension.load_inline, it maybe save build time: https://github.com/xuhancn/x86_isa_help?tab=readme-ov-file#build-speed-compared-between-ctypes-and-cpp_extension-load 
2. Move VecISA to separate python file.

Further work:
1. Enable Windows inductor.
2. Enable Windows inductor UTs.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

Differential Revision: [D52328687](https://our.internmc.facebook.com/intern/diff/D52328687)

### CODE REVIEW

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52328687).

File: torch/_inductor/jit_builder/cpp_builder.py [hunk 1]
```diff
@@ -0,0 +1,1022 @@
...
+            )
+    return cxx_path
+
+
+def _get_cpp_compiler() -> str:
+    if _IS_WINDOWS:
+        compiler = os.environ.get("CXX", "cl")
+    else:
+        if config.is_fbcode():
+            compiler = build_paths.cc()
```
Comment (desertfire) [inline torch/_inductor/jit_builder/cpp_builder.py:115]: -> `return build_paths.cc()`

File: torch/_inductor/jit_builder/cpp_builder.py [hunk 2]
```diff
@@ -0,0 +1,1022 @@
...
+        self._output_dir = ""
+        self._target_file = ""
+
+        self._name = name
+
+        if config.is_fbcode():
+            if BuildOption.get_aot_mode() and use_absolute_path:
+                # We need to copy any absolute-path torch includes
+                inp_name = [os.path.basename(i) for i in sources]
+                self._sources_args = " ".join(inp_name)
```
Comment (desertfire) [inline torch/_inductor/jit_builder/cpp_builder.py:884]: Missing `self._sources_args` in the else branch

**Review by desertfire — CHANGES_REQUESTED** [review]

File: torch/_inductor/jit_builder/cpp_builder.py (hunk 1 above)
Comment (xuhancn) [reply torch/_inductor/jit_builder/cpp_builder.py]: Done:

File: torch/_inductor/jit_builder/cpp_builder.py (hunk 2 above)
Comment (xuhancn) [reply torch/_inductor/jit_builder/cpp_builder.py:884]: Tried to fixed.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52328687).

**Pull request closed without merging**